Support APIServerAccessProfile in AzureManagedControlPlane

richardchen331 opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adds APIServerAccessProfile support in AzureManagedControlPlane. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1605

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [x] squashed commits
- [x] includes documentation
- [x] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Allow specifying APIServerAccessProfile in AzureManagedControlPlane.
```

mboersma commented:
/ok-to-test

richardchen331 commented:
/test pull-cluster-api-provider-azure-e2e-exp

CecileRobertMichon commented:
/assign @alexeldeib @LochanRn @nprokopic

nprokopic commented on exp/api/v1alpha3/azuremanagedcontrolplane_types.go:
```diff
@@ -94,6 +98,23 @@ type AADProfile struct {
+	AuthorizedIPRanges []string `json:"authorizedIPRanges,omitempty"`
+	// EnablePrivateCluster - Whether to create the cluster as a private cluster or not.
+	// +optional
+	EnablePrivateCluster bool `json:"enablePrivateCluster,omitempty"`
```
If this field is optional, I believe a bool pointer `*bool` would be better. Same for `EnablePrivateClusterPublicFQDN`. WDYT?

richardchen331 replied:
@nprokopic Thanks for your feedback! I'm pretty new to golang and cluster api, and my (limited) understanding of using a pointer is that we could differentiate between a zero value and a not specified value, but then we have to deal with `nil`. For `EnablePrivateCluster`, I think explicitly passing in a zero value (`false`) is equivalent to not passing it (both tells Azure to not enable private cluster). Therefore I'm not aware of the benefit of using a pointer.

Could you help explain a bit more why a bool pointer is better? I think I must have missed something and I'd love to learn :) Thank you!

CecileRobertMichon replied:
It is part of the k8s API conventions:

> Therefore, we ask that pointers always be used with optional fields that do not have a built-in nil value.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

richardchen331 replied:
Got it, thanks for the reference! I'll change the fields to use pointers.

nprokopic replied:
> It is part of the k8s API conventions

I vaguely remembered something like that, and that we had a similar discussion for some other CAPZ type some time ago, just couldn't find it.

Thanks Cecile for referencing this.

LochanRn commented:
/test pull-cluster-api-provider-azure-e2e-windows

nprokopic commented on exp/api/v1alpha4/azuremanagedcontrolplane_default.go:
```diff
@@ -73,3 +73,10 @@ func (r *AzureManagedControlPlane) setDefaultSubnet() {
+
+// setDefaultAPIServerAccessProfile sets the default APIServerAccessProfile for an AzureManagedControlPlane.
+func (r *AzureManagedControlPlane) setDefaultAPIServerAccessProfile() {
+	if r.Spec.APIServerAccessProfile == nil {
```
nit: Is this required? Wouldn't just leaving `nil` also work?

@CecileRobertMichon are there some other examples of setting default values like this?

richardchen331 replied:
You are right, I don't think this is required. (I was thinking setting it so there will be less working in validation webhook, but i don't think this is a good idea, since it's not good for unit test)

nprokopic commented on exp/api/v1alpha4/azuremanagedcontrolplane_webhook.go:
```diff
@@ -306,3 +312,71 @@ func (r *AzureManagedControlPlane) validateSSHKey() error {
+}
+
+// validateAPIServerAccessProfileUpdate validates update to APIServerAccessProfile.
+func (r *AzureManagedControlPlane) validateAPIServerAccessProfileUpdate(old *AzureManagedControlPlane) field.ErrorList {
```
There are quite a few field combinations here, it would be good to have unit tests for this.

Also, maybe using `DeepEqual` would make this shorter, like here for example https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/v0.5.2/api/v1alpha4/azurecluster_webhook.go#L109. @alexeldeib @CecileRobertMichon is using `DeepEqual` for comparison recommended/acceptable, or is handpicked field comparison like here more preferable?

richardchen331 replied:
Sure, I'll add unit tests for this.

Unfortunately I cannot use `DeepEqual` here since one of the field (`AuthorizedIPRanges`) is mutable. One option is to move `AuthorizedIPRanges` out of `APIServerAccessProfile` so we can use `DeepEqual` here. I don't have a strong preference here. Would love to know what do people think.

nprokopic replied:
You could normalize values before comparison, i.e. set mutable fields to empty/`nil` values in both operands (you would have to copy them first and compare normalized copies to avoid operand mutation).

No strong opinion for using `DeepEqual`, just a suggestion.

richardchen331 replied:
Thanks for the suggestion! I like that, it makes the code much simpler.

The downside is that it will prevent certain updates that is validate (e.g. change `EnablePrivateCluster` from `nil` to a pointer to `false`).

I've updated the code to use `DeepEqual`, but I'm open to change it back if people have strong opinions.

LochanRn replied:
Just a suggestion, one way to avoid DeepEqual while comparing is use cmp.Diff from github.com/google/go-cmp/cmp.

alexeldeib replied:
cmp uses reflection quite a bit, it just wraps some of the tricky bits of comparisons nicely.

LochanRn replied:
ohh okay

richardchen331 commented:
/retest

richardchen331 commented:
/retest

LochanRn commented on exp/api/v1alpha4/azuremanagedcontrolplane_webhook_test.go:
```diff
@@ -606,6 +620,44 @@ func TestAzureManagedControlPlane_ValidateUpdate(t *testing.T) {
+					DNSServiceIP: to.StringPtr("192.168.0.0"),
+					Version:      "v1.18.0",
+					APIServerAccessProfile: &APIServerAccessProfile{
+						EnablePrivateCluster: &boolTrue,
```
use EnablePrivateCluster: to.BoolPtr(true),

LochanRn commented:
Please squash all your commits into a single commit once you are done with all your changes

richardchen331 commented:
> Please squash all your commits into a single commit once you are done with all your changes

@LochanRn  Thanks for the feedback! I updated the code to use `to.BoolPtr(true)` and squashed all commits.

alexeldeib commented:
/approve

nice work!

alexeldeib commented:
will proactively lgtm since @LochanRn already did but got cleared 🙂 

/lgtm

alexeldeib commented:
/hold

alexeldeib commented on exp/api/v1alpha3/azuremanagedcontrolplane_types.go:
```diff
@@ -94,6 +98,23 @@ type AADProfile struct {
 }
 
+// APIServerAccessProfile - access profile for AKS API server.
+type APIServerAccessProfile struct {
```
similar comment to https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1679/files#r706664083

I think you should add this for v1a4 only, and implement the conversion webhook? which should basically be a noop except storing the data to roundtrip v1a4 -> v1a3 -> v1a4 conversion

richardchen331 replied:
Thanks @alexeldeib ! Reverted change in v1a3 and added one line in `azuremanagedcontrolplane_conversion.go` to restore the field during conversion.

alexeldeib commented:
/approve cancel
/lgtm cancel
/hold cancel

richardchen331 commented:
/retest

LochanRn commented:
@richardchen331  can you please also update the managedclusters documentation about APIServerAccessProfile. 

@alexeldeib over all this PR looks fine to me now.  Over to you for a final review.

richardchen331 commented:
> @richardchen331 can you please also update the managedclusters documentation about APIServerAccessProfile.
> 
> @alexeldeib over all this PR looks fine to me now. Over to you for a final review.

Sure. I'll rebase and update the documentation after the other open PR is merged since they modify similar set of files :)

richardchen331 commented:
@alexeldeib I just rebased this PR and also added documentation. Could you help take a final look? Thanks!

richardchen331 commented:
/retest

richardchen331 commented:
/retest

richardchen331 commented:
/retest

richardchen331 commented:
/retest

alexeldeib commented on azure/types.go:
```diff
@@ -407,6 +410,18 @@ type LoadBalancerProfile struct {
+	// PrivateDNSZone - Private dns zone mode for private cluster.
+	PrivateDNSZone *string
+	// EnablePrivateClusterPublicFQDN - Whether to create additional public FQDN for private cluster or not.
+	EnablePrivateClusterPublicFQDN *bool
```
I believe this will default to true on AKS side if private cluster is enabled. should we let that happen, or default it inside CAPZ?

it would only mean that a user who explicitly sets it would trigger a no-op reconcile, and a user who wants to change it will be able to.

richardchen331 replied:
Thanks for the feedback! I would slightly prefer letting AKS to handle the default. My rationales are:
1. This is true for other fields as well (e.g. [managedOutboundIPs](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/exp/api/v1alpha4/azuremanagedcontrolplane_types.go#L127), [sku tier](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/exp/api/v1alpha4/azuremanagedcontrolplane_types.go#L115))
2. We need to add some logic when choosing the default values (e.g. set `EnablePrivateClusterPublicFQDN` to true when `EnablePrivateCluster` is true)
3. Given 1,2, this will add some extra complexity to the CAPZ code
The downside is that there will be some unnecessary no-op reconciles, but i feel that's probably fine?

I'm open to change it if you feel strongly about it :)

alexeldeib replied:
Fine w me, I asked because I know this defaulting too true has surprised some folks in the past

alexeldeib replied:
But that’s really an aks issue

alexeldeib commented:
/test pull-cluster-api-provider-azure-e2e-exp 

probably related to the failures we've been discussing. I chatted a bit with @LochanRn about watch issues unrelated to the AKS backend issue, we still have some debugging to do

alexeldeib commented:
/approve
/lgtm

(adding both since this was already approved)

one minor comment but not a blocker

richardchen331 commented:
/retest

alexeldeib commented:
/lgtm
